jvm/sargon-android/src/main/java/com/radixdlt/sargon/os/driver/AndroidStorageDriver.kt

```diff
@@ -105,12 +88,58 @@ internal class AndroidStorageDriver(
         Result.success(mapping)
     }
 
-    private fun <T> Result<T>.reportFailure(message: String, commonError: CommonException) =
-        onFailure { error ->
-            Timber.tag("Sargon").w(error, message)
-            when (error) {
-                is CommonException -> throw error
-                else -> throw commonError
-            }
+    private fun <T> Result<T>.reportSecureStorageReadFailure(
+        key: SecureStorageKey
+    ) = onFailure { error ->
+        throw when (error) {
+            is BiometricsFailure -> error.toCommonException(key)
+            is CommonException -> error
+            else -> CommonException.SecureStorageReadException()
         }
+    }
+
+    private fun <T> Result<T>.reportSecureStorageWriteFailure(
+        key: SecureStorageKey
+    ) = onFailure { error ->
+        throw when (error) {
+            is BiometricsFailure -> error.toCommonException(key)
+            is CommonException -> error
+            else -> CommonException.SecureStorageWriteException()
+        }
+    }
+
+    private fun <T> Result<T>.reportUnsafeStorageReadFailure() = onFailure { error ->
+        throw when (error) {
+            is CommonException -> error
+            else -> CommonException.UnsafeStorageReadException()
+        }
+    }
+
+    private fun <T> Result<T>.reportUnsafeStorageWriteFailure() = onFailure { error ->
+        throw when (error) {
+            is CommonException -> error
+            else -> CommonException.UnsafeStorageWriteException()
+        }
+    }
+
+    private fun BiometricsFailure.toCommonException(key: SecureStorageKey) = CommonException.SecureStorageAccessException(
+        key = key,
+        errorKind = when (errorCode) {
+            BiometricPrompt.ERROR_CANCELED -> SecureStorageAccessErrorKind.CANCELLED
+            BiometricPrompt.ERROR_HW_NOT_PRESENT -> SecureStorageAccessErrorKind.HARDWARE_NOT_PRESENT
+            BiometricPrompt.ERROR_HW_UNAVAILABLE -> SecureStorageAccessErrorKind.HARDWARE_UNAVAILABLE
+            BiometricPrompt.ERROR_LOCKOUT -> SecureStorageAccessErrorKind.LOCKOUT
+            BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> SecureStorageAccessErrorKind.LOCKOUT_PERMANENT
+            BiometricPrompt.ERROR_NEGATIVE_BUTTON -> SecureStorageAccessErrorKind.NEGATIVE_BUTTON
+            BiometricPrompt.ERROR_NO_BIOMETRICS -> SecureStorageAccessErrorKind.NO_BIOMETRICS
+            BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL -> SecureStorageAccessErrorKind.NO_DEVICE_CREDENTIAL
+            BiometricPrompt.ERROR_NO_SPACE -> SecureStorageAccessErrorKind.NO_SPACE
+            BiometricPrompt.ERROR_TIMEOUT -> SecureStorageAccessErrorKind.TIMEOUT
+            BiometricPrompt.ERROR_UNABLE_TO_PROCESS -> SecureStorageAccessErrorKind.UNABLE_TO_PROCESS
+            BiometricPrompt.ERROR_USER_CANCELED -> SecureStorageAccessErrorKind.USER_CANCELLED
+            BiometricPrompt.ERROR_VENDOR -> SecureStorageAccessErrorKind.VENDOR
```

Comment:
Do we actually care to be so granular? or some grouping can be made?

Reply:
I would like to keep those in order to be able to debug all problems related to biometrics and keystore. These values pass from android driver through ffi to sargon and back to host.